Integrate Amplitude parachain into apps

ebma opened the pull request:
This PR is supposed to add Amplitude to the supported networks in swap and pool.

- [x] Adds implementations for amplitude to `parachains-impl`. Most of bifrosts components are re-used, but the balances required custom logic for the currency ID and conversion.
- [x] Adds the parachain ID of amplitude to several configs
- [x] Conditionally show amplitude components in the `/compat/` components 
- [x] Change the `ApiProvider` implementation. This was necessary because with the current implementation, the `registsry` instance in `packages/polkadot/typeRegistry.ts` is re-used for the bifrost and amplitude API instance, which is problematic and leads to issues with type conversions (especially the CurrencyId)
- [x] Adds a token list for amplitude and some icons


We currently do not have a subscan integration for Amplitude so the link that users get redirected to when clicking on "View Details" is not correct. But since using another block explorer is not supported at the moment, I kept it that way.

## Notes for testing
There is no pair or farming pool on Amplitude yet. We want to time the creation of those with the deployment of the Amplitude network into the Zenlink UI, so these changes need to be approved before we create the pools on-chain. However, the changes of this PR can be tested on our test network (Foucoco). The pair and farming pool that will be created on Amplitude are already set up on Foucoco. In order to test this, you need to:
- Change `rpc` for Amplitude in `packages/chain/parachains.json` to `wss://pencol-roa-00.pendulumchain.tech`
- Change `rpc` for Amplitude in `packages/config/polkadot/src/index.ts` to `wss://pencol-roa-00.pendulumchain.tech`
- Change `SQUID_HOST` for Amplitude in `packages/config/graph/src/index.ts` to`${SQUID_HOST_ENDPOINT}/foucoco-squid/graphql`
- Change `ARCHIVE_HOST` for Amplitude in `packages/config/graph/src/index.ts` to`https://foucoco.explorer.subsquid.io/graphql`

You can use the following secret mnemonic phase to import an account that has the required tokens. 
```
aisle mix lunar urge chair theme deny faint ensure jeans cream iron
``` 
You should use the farming pool with PID 2 for testing that the farming feature works.

## Icon assets for tokens
The assets that are supposed to be used for the Amplitude tokens are added in [this](https://github.com/zenlinkpro/assets/pull/69) PR. It still needs to be reviewed.

<!-- start pr-codex -->

---

## PR-Codex overview
This PR focuses on adding support for the Amplitude parachain. 

### Detailed summary
- Added exports for currency, address, useBalance, and useFarmBalance in the amplitude library.
- Added eslint configuration for amplitude.
- Added types for FarmBalanceMap and BalanceMap in the amplitude library.
- Added exports for systems and utils in the amplitude library.
- Updated the configuration files to include Amplitude in various lists and arrays.
- Added SwapRouter, components, hooks, systems, and types exports in the amplitude library.
- Added tsconfig.json file for amplitude.
- Updated getCurrencyCombinations.ts in the currency package.
- Updated config.ts in the token-price API.
- Updated package.json files in the zenlink-stats and polkadot packages.
- Updated tokens.ts in the currency package.
- Updated index.ts in the amplitude components.
- Updated types.ts in the amplitude library.
- Updated usePairs.ts in the amplitude hooks.
- Updated package.json file in the bifrost package.
- Updated useStablePools.ts in the amplitude hooks.
- Added amplitude.json file in the token-lists package.
- Updated parachains.json file in the chain package.
- Updated config.ts in the zenlink-stats API.
- Updated useTokensFromPool.ts in the shared package.
- Updated index.ts in the amplitude hooks.
- Updated useTokensFromPair.ts in the pool app.
- Updated tokenAddresses.ts in the currency package.
- Updated package.json file in the compat package.
- Updated client.ts in the graph-client package.
- Updated TokenApproveButton.tsx in the compat systems.
- Updated index.tsx in the swap app.

> The following files were skipped due to too many changes: `apps/swap/pages/index.tsx`, `packages/amm/src/Pair/Pair.ts`, `packages/parachains-impl/amplitude/libs/constants/address.ts`, `packages/compat/systems/Checker/Network.tsx`, `packages/ui/icons/network/circle/index.tsx`, `packages/compat/systems/Checker/Connected.tsx`, `packages/ui/icons/network/naked/index.tsx`, `packages/config/polkadot/src/index.ts`, `packages/chain/index.ts`, `packages/config/graph/src/index.ts`, `packages/parachains-impl/bifrost/libs/formats/currency.ts`, `packages/polkadot/package.json`, `packages/compat/hooks/useFarmRewards.ts`, `packages/config/router/src/index.ts`, `packages/compat/hooks/useWithdrawFarmingReview.ts`, `packages/compat/hooks/useSwapReview.ts`, `packages/compat/hooks/usePairs.ts`, `packages/compat/systems/Checker/Amounts.tsx`, `packages/ui/icons/network/naked/AmplitudeNaked.tsx`, `packages/token-lists/lists/index.ts`, `packages/compat/components/TokenSelector/TokenListImportChecker.tsx`, `packages/ui/currency/Icon.tsx`, `packages/ui/icons/network/circle/AmplitudeCircle.tsx`, `packages/compat/components/TokenListFilterByQuery/TokenListFilterByQuery.tsx`, `packages/compat/hooks/useStakeLiquidityStableReview.ts`, `packages/compat/hooks/useStablePools.ts`, `packages/compat/hooks/useClaimFarmingRewardsReview.ts`, `packages/parachains-impl/amplitude/hooks/usePairTotalSupply.ts`, `packages/compat/hooks/usePairTotalSupply.ts`, `packages/compat/hooks/useBalance/useBalance.ts`, `packages/compat/hooks/useStakeLiquidityStandardReview.ts`, `packages/compat/hooks/useAddLiquidityStandardReview.ts`, `packages/compat/hooks/useRemoveLiquidityStandardReview.ts`, `packages/compat/components/Profile/Profile.tsx`, `packages/parachains-impl/amplitude/SwapRouter.ts`, `packages/parachains-impl/amplitude/package.json`, `packages/parachains-impl/amplitude/hooks/useClaimFarmingRewardsReview.ts`, `packages/parachains-impl/amplitude/hooks/useTotalSupply.ts`, `packages/parachains-impl/amplitude/hooks/useStakeLiquidityStandardReview.ts`, `packages/parachains-impl/amplitude/hooks/useWithdrawFarmingReview.ts`, `packages/polkadot/components/ApiProvider.tsx`, `packages/parachains-impl/amplitude/hooks/useFarmBalance/useFarmBalance.ts`, `packages/parachains-impl/amplitude/libs/formats/currency.ts`, `packages/parachains-impl/amplitude/hooks/useRemoveLiquidityStandardReview.ts`, `packages/parachains-impl/amplitude/hooks/useSwapReview.ts`, `packages/parachains-impl/amplitude/hooks/useBalance/useBalance.ts`, `packages/parachains-impl/amplitude/hooks/useAddLiquidityStandardReview.ts`, `packages/parachains-impl/amplitude/hooks/usePairs.ts`, `packages/parachains-impl/amplitude/hooks/useFarmRewards.ts`, `pnpm-lock.yaml`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

ebma commented:
The console shows some error when I select the `KSM` asset. Maybe I forgot some crucial configuration for that?

<img width="1726" alt="image" src="https://user-images.githubusercontent.com/6690623/230424103-578e4ebe-2b8c-494d-be00-d4e0915d71c8.png">

haike0513 commented:
> This PR is supposed to add Amplitude to the supported networks in swap and pool. To do this, I looked at the changes for including the Bifrost chain and added changes similar to that. If I forgot something or added too much please let me know and I will change it accordingly.
> 
> Note that `pnpm run build` fails with
> 
> ```
> app:build: warn  - You have enabled experimental feature (esmExternals) in next.config.mjs.
> app:build: warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/swap","destination":"undefined/swap"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/swap/:path*","destination":"undefined/swap/:path*"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/pool","destination":"undefined/pool"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/pool/:path*","destination":"undefined/pool/:path*"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/referrals","destination":"undefined/referrals"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/referrals/:path*","destination":"undefined/referrals/:path*"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/analytics","destination":"undefined/analytics"}
> app:build: 
> app:build: `destination` does not start with `/`, `http://`, or `https://` for route {"source":"/analytics/:path*","destination":"undefined/analytics/:path*"}
> app:build: 
> app:build: 
> app:build: Error: Invalid rewrites found
> ```
> 
> but this also happens to me on the latest main branch.

You should add the `env` file on apps package, You can refer the file https://github.com/zenlinkpro/zenlink-interface/blob/main/apps/app/.env.example

haike0513 commented:
> The console shows some error when I select the `KSM` asset. Maybe I forgot some crucial configuration for that?
> 
> <img alt="image" width="1726" src="https://user-images.githubusercontent.com/6690623/230424103-578e4ebe-2b8c-494d-be00-d4e0915d71c8.png">

Please read the https://github.com/zenlinkpro/zenlink-interface/blob/main/docs/Integration.md ,You should complete `Basic operation implementation` and  `Add GraphQL Service` at first

ebma commented:
Thanks for the help @haike0513 , we are working on the graphQL service for the time being and I will update the PR once it's ready.

ebma commented:
@haike0513 I reworked the implementation and fixed the problems with `pnpm run build`. I edited the original description of this PR to give more info on the changes. Could you please have a look?

haike0513 commented:
> @haike0513 I reworked the implementation and fixed the problems with `pnpm run build`. I edited the original description of this PR to give more info on the changes. Could you please have a look?

Do you have a testnet about this, there is no pairs on your mainnet.It's better to have a testnet, so that I can test is it works well

ebma commented:
Hey @haike0513, I now updated the PR and fixed the remaining issues, so it is ready for a proper review now. You can now test the changes on our test network (Foucoco). I added the necessary steps for doing so to the description of this PR. Let me know if you need more information.

haike0513 commented on packages/compat/hooks/useBalance/useBalance.ts:
```diff
@@ -41,6 +42,14 @@ export const useBalances: UseBalances = ({
     chainId,
     account,
     currencies,
+    enabled: chainId === ParachainId.BIFROST_KUSAMA,
```
Bifrost has two chain on Kusama and Polkadot, this only enable in kusama

haike0513 commented on packages/parachains-impl/amplitude/hooks/index.ts:
```diff
@@ -0,0 +1,13 @@
+export * from '../../bifrost/hooks/useStablePools'
+export * from '../../bifrost/hooks/useStakeLiquidityStandardReview'
+export * from '../../bifrost/hooks/useSwapReview'
+export * from '../../bifrost/hooks/useTotalSupply'
```
This is just reexport the bifrost impl, If there is a break upgrade on bifrost, it maybe not work on your chain. So It's better you can have your own impl.

ebma replied:
Right, that's a good point. I changed it so that the hooks for Amplitude don't reference the ones of Bifrost anymore.
However, I kept the [components](https://github.com/pendulum-chain/zenlink-interface/blob/55125c20b198bd58ecb68aec0fb4ca7927b3584c/packages/parachains-impl/amplitude/components/index.ts) as-is meaning that the components for Amplitude will still be the same as for Bifrost. I decided not to change that to reduce maintenance overhead for these components. Is that okay @haike0513 or would you also suggest to duplicate those?

haike0513 commented on packages/parachains-impl/amplitude/hooks/useBalance/useBalance.ts:
```diff
@@ -0,0 +1,119 @@
+      else
+        result[validatedTokens[i].address] = Amount.fromRawAmount(validatedTokens[i], '0')
+
+      // BNC
```
May be you can remove this line

haike0513 commented on packages/parachains-impl/bifrost/libs/formats/currency.ts:
```diff
@@ -1,7 +1,9 @@
+import { ParachainId } from '@zenlink-interface/chain'
 import type { NodePrimitivesCurrency } from '../../types'
 import { pairAddressToAssets } from '../constants'
+import { parseNodePrimitivesCurrency as amplitudeParseNodePrimitivesCurrency } from '../../../amplitude/libs/formats/currency'
```
It's better to impl it on your own chain impl, not include in another chain impl

ebma replied:
As far as I can tell this is necessary, because, let's say you are connected to the 'Amplitude' network in the UI and try to navigate to the details for a Bifrost pool. You will encounter the following error because the app will try to parse the currencies with the Amplitude implementation although it's a bifrost pool and should use the bifrost parsing instead. The same is necessary vice versa. That's why I added [this](https://github.com/pendulum-chain/zenlink-interface/blob/7b6dc630ecd0516ebb04f02341da18dbe7dc654c/packages/parachains-impl/amplitude/libs/formats/currency.ts#L44-L45) and [this](https://github.com/pendulum-chain/zenlink-interface/blob/7b6dc630ecd0516ebb04f02341da18dbe7dc654c/packages/parachains-impl/bifrost/libs/formats/currency.ts#L92-L93) statement.

<img width="1144" alt="image" src="https://github.com/zenlinkpro/zenlink-interface/assets/6690623/42aac698-c2a3-4109-9912-7f407bcb76f9">

haike0513 commented on packages/wagmi/hooks/usePairTotalSupply.ts:
```diff
@@ -21,7 +20,7 @@ export const usePairTotalSupply = (pair: Pair | undefined | null, chainId: Parac
         address,
         name: 'Zenlink LP Token',
-        decimals: 18,
+        decimals: chainId === ParachainId.AMPLITUDE ? 12 : 18,
```
I think it's no need to add this, the hooks in wagmi package is about evm network

ebma commented:
@haike0513 thanks for your valuable feedback on the review. I addressed all your comments, let me know if this is now ready for merge or if there still are some other issues. 
Please **don't merge** this PR just yet though, as we'd like to have a specific timing about when these changes would go live on the production Zenlink UI. It's important to know whether this PR is overall in a mergeable state though.

haike0513 commented:
Can you switch the `rpc`, `SQUID_HOST`, `ARCHIVE_HOST` to your  test network (Foucoco), So we can test it on the test network in Vercel preview mode. When the PR is ready we can merge it with disable `Amplitude` by the config. And when the Amplitude is ready for production, you can open another PR to enable it.

ebma commented:
@haike0513 I changed the parameters so that it now uses Foucoco. You should now be able to test it.

> When the PR is ready we can merge it with disable Amplitude by the config. And when the Amplitude is ready for production, you can open another PR to enable it.

Sounds great, thank you 🙏
